compiler/rustc_borrowck/src/region_infer/opaque_types.rs

```diff
@@ -227,6 +227,17 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 opaque_type_key,
                 universal_concrete_type,
             );
+
+            // Sometimes, when the hidden type is an inference variable, it can happen that
+            // the hidden type becomes the opaque type itself. In this case, this was an opaque
+            // usage of the opaque type and we can ignore it. This check is mirrored in typeck's
+            // writeback.
```

Comment:
```suggestion
            // FIXME(-Znext-solver): This should be unnecessary with the new solver.
```
Also, can this not result in an overflow? If you have `Opaque = OPaque` in the storage and then equate it with `u32`.

It tries to assign `u32` to `Opaque`, which fetches the hidden type of `Opaque` which is `OPaque` again, causing the equate of the previous hidden type with `u32` to use `handle_opaque_type` again?

Reply:
By taking a glance, I guess it makes this fixes the first example of #115017. Please add it as a test.
It is interesting to see how it interacts with this issue as a whole.

Reply:
The first example already works on master